### PR TITLE
Fix audio tracks not looping

### PR DESCRIPTION
**Pull request by Aergwyn**
This fixes songs not looping because they got cleaned up before they had a chance to start anew.
Solves https://github.com/ppy/osu/issues/1522

### CODE REVIEW

File: osu.Framework/Audio/Track/TrackBass.cs [hunk 1]
```diff
@@ -233,7 +233,7 @@ internal override void OnStateChanged()
 
         public override int? Bitrate => bitrate;
 
-        public override bool HasCompleted => base.HasCompleted || IsLoaded && !IsRunning && CurrentTime >= Length;
+        public override bool HasCompleted => base.HasCompleted || HasReachedEnd && !Looping;
```
Comment (peppy) [inline osu.Framework/Audio/Track/TrackBass.cs]: why isn't this implemented in `Track`?

Comment (Aergwyn) [reply osu.Framework/Audio/Track/TrackBass.cs]: I guess my idea was to not having everything loopable by default. But given that `Looping` is in `Track` anyways it doesn't make much sense. I'll move it up then.

**Comment by Aergwyn** [comment]
@peppy is this ok now?

**Comment by peppy** [comment]
Have not yet had a chance to review this. No need to highlight me; I (or someone else) will get to it as time allows

File: osu.Framework/Audio/Track/TrackVirtual.cs [hunk 2]
```diff
@@ -66,7 +66,7 @@ public override bool HasCompleted
         {
             get
             {
-                lock (clock) return base.HasCompleted || IsLoaded && !IsRunning && CurrentTime >= Length;
+                lock (clock) return base.HasCompleted;
```
Comment (peppy) [inline osu.Framework/Audio/Track/TrackVirtual.cs]: do we even need this override? i think it's safe without this clock-lock

File: osu.Framework/Audio/Track/Track.cs [hunk 3]
```diff
@@ -101,10 +121,9 @@ protected override void UpdateState()
         {
             FrameStatistics.Increment(StatisticsCounterType.Tracks);
 
-            if (Looping && HasCompleted)
+            if (Looping && HasReachedEnd)
             {
```
Comment (peppy) [inline osu.Framework/Audio/Track/Track.cs]: minor but these braces are no longer needed

File: osu.Framework/Audio/Track/Track.cs [hunk 4]
```diff
@@ -90,6 +103,13 @@ public virtual double Rate
 
         public bool IsReversed => Rate < 0;
 
+        public override bool HasCompleted => base.HasCompleted || HasReachedEnd && !Looping;
+
+        /// <summary>
+        /// Stats if this track has reached its end and isn't running.
```
Comment (peppy) [inline osu.Framework/Audio/Track/Track.cs]: Stats -> States

or just reword to "Whether this track has..."

**Pull request merged**